### PR TITLE
search: Add new search query input proof of concept

### PR DESCRIPTION
**Pull request by fkling**
This PR adds a proof of concept implementation for a new search query input.

https://user-images.githubusercontent.com/179026/206053660-6fc7d41d-8573-4578-8db1-f4ba0910ccb3.mp4

When enabled (via `experimentalFeatures.searchQueryInput: "experimental"` (up for debate)), the search query input on the *search home page* is replaced with the experimental implementation.

I "contained" the changes as much as possible to `client/search-ui/src/input/experimental`, but I also had to move some code around for reuse.

The experimental implementation uses Svelte for easy prototyping and  is loaded lazily so it should not impact any production builds.

NOTE: I'm tagging other engineers mostly for awareness and high level feedback, not for detailed code review.

Todo:

-  [x] Enable feature flag check (disabled to make local testing easier)

## Test plan

Manual testing.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.04% (+1.20 kb) | -0.07% (-10.73 kb) 🔽 | -0.10% (-11.93 kb) 🔽 | 1.10% (+8) 🔺 |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/fkling/exprimental/search-input/compare-report.html;tab=live_object) for a full comparison between the commits 7b05eb7 and 79436af or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: tsconfig.json [hunk 1]
```diff
@@ -18,14 +18,19 @@
     "composite": true,
     "outDir": "out",
     "rootDir": ".",
+    "plugins": [
+      {
+        "name": "typescript-svelte-plugin"
```
Comment (valerybugakov) [inline tsconfig.json:23]: Hey, @fkling, I think the prototype looks fantastic, and I love the direction where we're going with it! I'm worried about the technical shift behind this PR. Bringing another experimental language for easy prototyping without discussing it properly in the RFC, where we can align on the future direction, is disturbing. We kind of touched on the topic of rewriting the web application with the new technology (particularly with Svelte) in [the Rewrite the web app RFC](https://docs.google.com/document/d/1jEf5f04HKMT0ieG7HuvFgJtyk6J9HXZaWU9tfGABYYA/edit) and my conclusion from it was that we agree that rewriting the application with the new tech won't solve our problems. We need to focus on better guidelines and the improved core architecture of the web app that will radically simplify things we've accumulated over years of product development, which currently has a lot of outdated practices that are slowing us down. That's where we could invest our extra work cycles to get the best ROI.

The feedback captured in [the RFC](https://docs.google.com/document/d/1jEf5f04HKMT0ieG7HuvFgJtyk6J9HXZaWU9tfGABYYA/edit) itself and in the Frontend Crew meeting where we discussed the rewrite topic doesn't signal that we still decided to go into "the rewrite the web app in Svelte" direction. Please, let me know if I missed any other discussions related to the topic. (I did a "Svelte" search in our Slack organization, but it yielded no relevant results.)

I do not want to duplicate the feedback points which are already present in the initial rewrite RFC. Still, I want to highlight three significant issues with the new experimental language for easy prototyping.

1. We already have one experimental language for experimental user interfaces — Elm. Nobody owns it right now, and it's not trivial to start using the functionality that Elm implements, even if we want to move the UIs built with it to production. It has multiple issues starting from the considerable bundle size to WCAG accessibility which does not automatically benefit from the Wildcard component library improvements we implemented this year. The same potential issues should be reviewed and discussed for a new language to avoid bumping into them in a critical moment when we decide to release experimental interfaces to customers.
2. Elm and other additional languages add infrastructure maintenance costs. Elm caused multiple problems with Webpack, esbuild, and CI configurations. We didn't have a chance to discuss these when Elm was introduced to the codebase. It was a surprise for the Frontend Platform team when we discovered issues related to it. We should not follow the same pattern with the new language. Keeping in mind the fact we have a bunch of critical problems with the current client infrastructure ([slow Typescript builds](https://sourcegraph.slack.com/archives/C01C3NCGD40/p1669859170510239), [slow Webpack builds](https://sourcegraph.slack.com/archives/C07KZF47K/p1668490432355359), an experimental esbuild loader, and experimental Bazel integration that's a work in progress), increasing our client build complexity without considering all the effects is dangerous and can bite us in the future. Let me know if we already discussed these somewhere.
3. We have [a documented process](https://docs.sourcegraph.com/dev/background-information/tech_stack#tech-stack) for changing our tech stack, and I think it's important to comply with it because it adds clarity to decisions made and makes the codebase predictable for everybody at the organization. 

Zooming in from the higher level question about adding a new language to the question of using the experimental language to test the water and see what other engineers think about it, it's essential not to invest into it heavily in building critical product features with it because later, if we want to shift to another approach, it's not trivial to do. After all, it's not a tiny side admin feature built with the experimental language; it's the core piece of the web application, the search input. This is a risk that we could easily mitigate by trying out new approaches on the sides to gauge ROI first, and then when we're sure that this is the path worth taking, commit to it in more critical parts of the product.

Let me know what you think!
And @sourcegraph/frontend-devs, I'm curious, what are your thoughts on the topic too?

**Comment by olafurpg** [comment]
> The experimental implementation uses Svelte for easy prototyping and is loaded lazily so it should not impact any production builds.

How difficult would it be to implement these UX improvements in React? The improvements from the video look fantastic and I'd love if we can turn this on by default with the January/February releases.

As discussed offline, I'm supportive of Svelte. I'm mostly concerned that mixing Svelte with this desperately needed UX improvement will delay the rollout the UX improvement.

**Comment by olafurpg** [comment]
How do I build this locally? Running `yarn` fails for me with link errors

```
❯ yarn
➤ YN0000: ┌ Resolution step
➤ YN0000: └ Completed in 0s 287ms
➤ YN0000: ┌ Fetch step
➤ YN0000: └ Completed in 0s 641ms
➤ YN0000: ┌ Link step
➤ YN0009: │ deasync@npm:0.1.28 couldn't be built successfully (exit code 1, logs can be found here: /private/var/folders/6l/bq1gqxk532564klg4cjhp50c0000gn/T/xfs-a924cb68/build.log)
➤ YN0000: └ Completed in 1s 685ms
➤ YN0000: Failed with errors in 2s 733ms
```

**Comment by fkling** [comment]
@olafurpg 

> How do I build this locally? Running yarn fails for me with link errors

@quinnkeast also had problems, maybe it's macOS specific. I switched to different package that doesn't use deasync. I hope that fixes it.

**Comment by fkling** [comment]
@valerybugakov, @olafurpg on Svelte:

I purposefully added as little business logic to UI components as possible, so that the UI components are easy to replace.

> my conclusion from it was that we agree that rewriting the application with the new tech won't solve our problems

My main conclusion was that we want to address the problems in a different way, but we didn't categorically exclude Svelte. I still think that it makes to try out Svelte where performance or other interop behavior is critical, and this is such a place:

- On any data changes only the minimal amount of work should be done to update the query input
- Integrating UI components with CodeMirror extensions is a bit more convenient with Svelte. There is no need to create new React roots etc (I don't even know what the implications of creating a new root is).

I'm happy to create a separate RFC for this.

> Elm and other experimental languages add infrastructure maintenance costs. Elm caused multiple problems with Webpack, esbuild, and CI configurations

Weren't there plans to remove elm at this point? But anyways, I can understand the worries. At the same time I believe Svelte is easier to integrate than Elm, so it's less of a risk.

But if others deem the risk of adding Svelte for the prototype to be too high then I have no problems using React. Though I do think it would be missed opportunity.

Also keep in mind this is still a proof of concept. It's possible that none of the code makes it into the final "feature".

**Comment by vovakulikov** [comment]
@fkling I 100% agree that this PR prototype looks and feels fantastic. I like the UX and the product part of this. I do have a few thoughts on the tech side too. 

All this story reminds me that in my previous company where I used to work (Wrike), we had the biggest codebase (after Google), with 2-3M LOC written in Dart. And most of the engs didn't like it because developing a web app on dart was very painful, it wasn't performant enough, but everyone was more or less ok with that because this was our past dependencies, and the codebase is huge. I got lucky working there with biased-to-action teammates, and we started a migration to React and Typescript. But before we started, we spent a lot of time figuring out a migration plan and tech corner cases, CI/CD processes. Because we knew that we had only one shot.

So why I'm telling you all of this? I think even if Svelte is a good framework, we can't start doing something big without a good understanding of all aspects and reasons for this change. I think Sourcegraph already has a lot of places where we've done just 80% of the work and then moved on to doing something else. I see this as the main reason for the situation where we are at today. 

- Lack of high-level architecture, 
- Basically, a bunch of different implementations of the same UI, 
- Different patterns and approaches
- Just some legacy code, things that were deprecated but still with us
- A ton of feature flags that somehow works, but nobody knows about all possible feature flag combinations 

This list of problems isn't React or any other technology specific. I'm 100% for bringing some ideas from different ecosystems but changing the tech stack won't solve our problems here. IMO. 

I'm open to the conversation, though. If you think I'm wrong in this statement, please let's have a pair session or Q&A session with other FE engs (or bring this to the FE crew meeting).

**Comment by quinnkeast** [comment]
>UPDATE: I also forgot to mention that, as @valerybugakov said above, the search box UI is the most core functionality of the product. We reuse search-ui heavily in the code insight and other parts of the product. Without a good understanding of how we would integrate it, it would be a huge blocker for us to adopt a new search box UI and align the search UI across the product. And by this, we would end up with two different implementations of the same UI.

Outside of the Svelte conversation, a quick response to this: in the [original RFC](https://docs.google.com/document/d/1eeb63tAEJ3hkJLh5_tTbwlUlfpnP8onJaBqi5_CWrAA/edit#bookmark=id.bwdtsmw1r311), we note that:

>For other places where the search input is shown (monitoring, insights, batch changes, IDE extensions), continue using the current search input for the time being. We’ll probably need to make an “embedded” variant on the search input to gracefully handle this distinct context (with different heuristics around what content is shown—e.g. you wouldn’t want a UI that lets you jump to matching symbols and exit a new entity creation flow).

I think there's probably going to need to be an explicit variant of the search input for these cases, because the rebooted search input is a different UI pattern overall than an input within a page. I don't think we need to look at this as a blocker at all.

File: client/search-ui/src/input/experimental/Suggestions.tsx [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    if (!show) {
+        return null
+    }
+
+    return (
+        <div
+            ref={setContainer}
+            id={id}
+            className={styles.suggestions}
+            role="grid"
```
Comment (felixfbecker) [inline client/search-ui/src/input/experimental/Suggestions.tsx:81]: What's the thinking behind using `grid`? Is it navigatable in 2 dimensions? I assumed it would be more something like a listbox or menu (whatever we had before)

**Comment by felixfbecker** [comment]
I love the UX in the video and am really looking forward to trying this!

File: client/search-ui/src/input/experimental/Suggestions.tsx (hunk 2 above)
Comment (fkling) [reply client/search-ui/src/input/experimental/Suggestions.tsx:81]: Nothing is set in stone of course, but my thinking was this: We show multiple pieces of information in a single row: label, description, etc. Apparently in the existing autocomplete popover every row is read in full, and so we added commas in between to improve how it's read.

When I looked into implementing the new solution I came across [this example](https://www.w3.org/WAI/ARIA/apg/example-index/combobox/grid-combo.html) where it's possible to navigate between different pieces of information via arrow keys. I liked it because I can also see a future where we might expose multiple actions per row and this would allow people to choose.

But again, I can totally be convinced to do it differently in the final product, it just has to fit.

**Comment by fkling** [comment]
Any (other) objections merging this so we can continue to iterate on this while collecting early feedback?

File: client/search-ui/src/input/experimental/Suggestions.tsx [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+                const [group, option] = match[0].split('x')
+                onSelect(results[+group].options[+option])
+            }
+        },
+        [onSelect, results]
+    )
+
+    const windowHeight = useWindowHeight()
+    const maxHeight = useMemo(
+        () => (container ? `${windowHeight - container.getBoundingClientRect().top - 20}px` : 'auto'),
```
Comment (valerybugakov) [inline client/search-ui/src/input/experimental/Suggestions.tsx:65]: Leaving a comment here explaining the `20px` number would be helpful.

File: client/search-ui/src/input/experimental/Suggestions.tsx [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                                    ? option.render(option)
+                                    : option.matches
+                                    ? [...option.value].map((char, index) =>
+                                          option.matches!.has(index) ? (
+                                              <span className={styles.match}>{char}</span>
+                                          ) : (
+                                              char
+                                          )
+                                      )
+                                    : option.value}
```
Comment (valerybugakov) [inline client/search-ui/src/input/experimental/Suggestions.tsx:115]: Looks similar to the `FilterOption` output. Maybe it makes sense to reuse it to keep them consistent.

File: client/search-ui/src/input/experimental/suggestions.ts [hunk 5]
```diff
@@ -0,0 +1,552 @@
...
+            ) {
+                const query = (this.running = new RunningQuery(source))
+                query.source
+                    .run()
+                    ?.then(result => {
+                        if (this.running === query) {
+                            this.view.dispatch({ effects: updateResultEffect.of({ source, result }) })
+                        }
+                    })
+                    .catch(() => {})
```
Comment (valerybugakov) [inline client/search-ui/src/input/experimental/suggestions.ts:175]: Would this error be handled somewhere else?

File: client/search-ui/src/input/experimental/suggestions.ts [hunk 6]
```diff
@@ -0,0 +1,552 @@
...
+        // input.
+        window.requestAnimationFrame(() => this.view.contentDOM.focus())
+    }
+
+    constructor(private readonly id: string, public readonly view: EditorView, public parent: HTMLDivElement) {
+        const state = view.state.field(suggestionsStateField)
+        this.container = document.createElement('div')
+        this.root = createRoot(this.container)
+        this.root.render(
+            React.createElement(Suggestions, {
```
Comment (valerybugakov) [inline client/search-ui/src/input/experimental/suggestions.ts:111]: Interesting! I know this is the experimental PR, and we probably don't want to overinvest in the documentation. Once it's on the way to being the default implementation, it would be awesome to have a high-level diagram about how the integration of React components with the CodeMirror works to build the mental model quickly

**Review by valerybugakov — APPROVED** [review]
Thanks, @fkling, for updating the internal implementation! Good thing that most of the stuff is handled by CodeMirror. I left a few optional comments in the review. Excited to start using this implementation behind the feature flag once it becomes available on S2!

File: client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.module.scss [hunk 7]
```diff
@@ -0,0 +1,79 @@
+.container {
+    flex: 1;
+    position: relative;
+
+    .spacer {
+        margin: 12px;
```
[the remark this answers is not in the document]
Comment (fkling) [reply client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.module.scss:6]: Didn't we want to ditch `rems` for margin and padding? Or maybe we haven't followed through with this yet.

File: client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.module.scss [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    button {
+        display: none;
+        align-self: flex-start;
+        padding: 0.125rem 0.25rem;
+        margin: 2px;
+        border: 0;
+        background-color: transparent;
+        border: 1px solid transparent;
```
[the remark this answers is not in the document]
Comment (fkling) [reply client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.module.scss:66]: Also I routinely forget to run style-lint before I push ;)

File: client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.module.scss [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        border-radius: 4px;
+
+        &:focus {
+            outline: 2px solid rgba(163, 208, 255, 1);
```
[the remark this answers is not in the document]
Comment (fkling) [reply client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.module.scss:70]: I'm happy to do this after we decide whether or not to move forward with the whole experiment.

File: client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.tsx [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+import { parseInputAsQuery } from '../codemirror/parsedQuery'
+import { filterHighlight, querySyntaxHighlighting } from '../codemirror/syntax-highlighting'
+
+import { editorConfigFacet, Source, suggestions } from './suggestions'
+
+import styles from './CodeMirrorQueryInputWrapper.module.scss'
+
+// Helper function to observe the current size of an element. This is used
+// to create an appropriately sized placeholder element.
+function useResizeObserver(node: HTMLElement | null): { height: number } {
```
[the remark this answers is not in the document]
Comment (fkling) [reply client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.tsx:24]: Didn't know that that existed :)

File: client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.tsx [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+export interface CodeMirrorQueryInputWrapperProps {
+    queryState: QueryState
+    onChange: (queryState: QueryState) => void
+    onSubmit: () => void
+    isLightTheme: boolean
+    interpretComments: boolean
+    patternType: SearchPatternType
+    placeholder: string
+    suggestionSource: Source
+    history: History
```
[the remark this answers is not in the document]
Comment (fkling) [reply client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.tsx:194]: Nope, that will do it too. I just used because it was already passed via props.

File: client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.tsx [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+    const hasValue = queryState.query.length > 0
+
+    return (
+        <div className={styles.container}>
+            {/* eslint-disable-next-line react/forbid-dom-props */}
+            <div className={styles.spacer} style={{ height: `${spacerHeight}px` }} />
+            <div className={styles.root}>
+                <div ref={setFocusContainer} className={styles.focusContainer}>
+                    {/* eslint-disable-next-line react/forbid-dom-props */}
+                    <div ref={setContainer} style={{ display: 'contents' }} />
```
[the remark this answers is not in the document]
Comment (fkling) [reply client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.tsx:281]: It would be nice if there was a utility class for this, just like we have for other `display` values.

File: client/search-ui/src/input/experimental/README.md [hunk 13]
```diff
@@ -0,0 +1,6 @@
+This folder contains a new (experimental) implementation of the search query
+input.
+
+IT SHOULD NOT BE USED!
+
+If you have questions about it reach out the search product team.
```
[the remark this answers is not in the document]
Comment (fkling) [reply client/search-ui/src/input/experimental/README.md:6]: You basically mean reverting what I did in https://github.com/sourcegraph/sourcegraph/pull/45310/files#diff-e2954b558f2aa82baff0e30964490d12942e0e251c1aa56c3294de6ec67b7cf5 ? I wanted to avoid the disabling the rule for the imports I'm adding but maybe it makes sense to do so in this case.

File: client/search-ui/src/input/experimental/suggestions.ts (hunk 5 above)
Comment (fkling) [reply client/search-ui/src/input/experimental/suggestions.ts:175]: No idea yet.

File: client/search-ui/src/input/experimental/suggestions.ts (hunk 6 above)
Comment (fkling) [reply client/search-ui/src/input/experimental/suggestions.ts:111]: FWIW we already do something similar for the CodeMirror blob view.

File: client/search-ui/src/input/experimental/Suggestions.tsx (hunk 4 above)
Comment (fkling) [reply client/search-ui/src/input/experimental/Suggestions.tsx:115]: Yeah, that should possibly be factored into yet another component, but I wanted to keep it simple for now.

File: client/search-ui/src/input/codemirror/syntax-highlighting.ts [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+]
+
+const validFilter = Decoration.mark({ class: 'sg-filter', inclusive: false })
+const invalidFilter = Decoration.mark({ class: 'sg-filter sg-invalid-filter', inclusive: false })
+
+export const filterHighlight = [
+    EditorView.baseTheme({
+        '.sg-filter': {
+            backgroundColor: 'var(--oc-blue-0)',
+            // border: '1px solid var(--oc-blue-1)',
```
Comment (limitedmage) [inline client/search-ui/src/input/codemirror/syntax-highlighting.ts:41]: nit: Can we remove this line?

**Review by limitedmage — COMMENTED** [review]
A few comments for an initial look:

1. Suggestions don't seem to be working for me? I get network events as I type but nothing is rendered.
 https://user-images.githubusercontent.com/206864/207167373-140849b1-29a9-42ee-a009-1e7d2fa28ba9.mp4
2. The dark mode filters seem too bright, I assume this will be addressed later as this feature doesn't have dark mode designs yet
 image

File: client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.module.scss (hunk 7 above)
[the remark this answers is not in the document]
Comment (valerybugakov) [reply client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.module.scss]: I remember we discussed it but never updated our stylelint config. Maybe we should do that in another PR 👍

File: client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.module.scss (hunk 9 above)
[the remark this answers is not in the document]
Comment (valerybugakov) [reply client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.module.scss:69]: Sounds good to me! Are we tracking things we want to fix if we decide to move forward later? This code probably won't go through another review cycle.

File: client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (valerybugakov) [reply client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.tsx:269]: Now, we can actually do that! All the global classes are in our codebase, not in the Bootstrap package. You can add one here: https://github.com/sourcegraph/sourcegraph/blob/1dbb1f6dbffae5dec374dbbabca2e49076c57cdc/client/branded/src/global-styles/utilities.scss#L41

File: client/search-ui/src/input/experimental/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (valerybugakov) [reply client/search-ui/src/input/experimental/README.md:6]: > You basically mean reverting what I did in ...

Yeah, we can add another group with your experimental folder and a unique message explaining the reason behind banning imports from it. Then the README will be obsolete because all the info will be communicated via the ESLint rule.

**Comment by fkling** [comment]
@limitedmage 
> Suggestions don't seem to be working for me? I get network events as I type but nothing is rendered.

That's weird... do you see any errors in the console? It works as expected for me.

File: client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.module.scss (hunk 9 above)
[the remark this answers is not in the document]
Comment (quinnkeast) [reply client/search-ui/src/input/experimental/CodeMirrorQueryInputWrapper.module.scss:69]: Yes, we need to track the specific things. @fkling and I will make a list of known things.

**Comment by quinnkeast** [comment]
> 2. The dark mode filters seem too bright, I assume this will be addressed later as this feature doesn't have dark mode designs yet
>    image

That's right, we haven't made the dark mode yet. We wanted to get the behaviours and interactions down before going too far into high-fidelity design polish.
